streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -1128,4 +1134,59 @@ object ZSink extends ZSinkPlatformSpecific {
       bucket  <- Ref.make((units, current)).toManaged_
     } yield bucketSink(bucket)
   }
+
+  /**
+   * Creates a sink which delays input elements of type A according to the given bandwidth parameters
+   * using the token bucket algorithm. The weight of each element is determined by the `costFn` function.
+   */
+  final def throttleShape[A](units: Long, duration: Duration)(
+    costFn: A => Long
+  ): ZManaged[Clock, Nothing, ZSink[Clock, Nothing, Nothing, A, A]] =
+    throttleShapeM[Any, Nothing, A](units, duration)(a => UIO.succeed(costFn(a)))
+
+  /**
+   * Creates a sink which delays input elements of type A according to the given bandwidth parameters
+   * using the token bucket algorithm. The weight of each element is determined by the `costFn` effectful
+   * function.
+   */
+  final def throttleShapeM[R, E, A](units: Long, duration: Duration)(
+    costFn: A => ZIO[R, E, Long]
+  ): ZManaged[R with Clock, E, ZSink[R with Clock, E, Nothing, A, A]] = {
+    import ZSink.internal._
+
+    def bucketSink(bucket: Ref[(Long, Long)]) = new ZSink[R with Clock, E, Nothing, A, A] {
+      type State = (Ref[(Long, Long)], Promise[E, A])
+
+      val initial = Promise.make[E, A].map(promise => Step.more((bucket, promise)))
+
+      def step(state: State, a: A) =
+        for {
+          weight  <- costFn(a)
+          current <- clock.currentTime(TimeUnit.NANOSECONDS)
+          delay <- state._1.modify {
+                    case (tokens, timestamp) =>
+                      val missing = weight - tokens
+                      val cycles =
+                        if (missing <= 0) 0
+                        else {
+                          val c = missing / units
+                          if (c * units < missing) c + 1 else c
+                        }
+                      val newTokens    = tokens + cycles * units - weight
+                      val newTimestamp = timestamp + cycles * duration.toNanos
+                      val delay        = Duration.Finite(newTimestamp - current)
+                      (delay, (newTokens, newTimestamp))
+                  }
+          _ <- state._2.succeed(a).delay(delay)
```

Comment:
`delay` forces a bounce through the platform's scheduler; can we skip it if `delay <= 0`? You can also do `clock.sleep(delay)` directly, no need to delay a value.

Reply:
Done.